File: app/src-tauri/src/beatify.rs
//! Beatify tab IPC (PRD "Beatify"): analyze a library track, inspect the
//! grid in the import modal, and commit a constant-tempo render.
//!
//! Like the Clip page this is a pure control-plane feature — it never
//! touches the engine or the RT thread, and every command is `async` so
//! the (multi-second) detection and warp render run on Tauri's worker pool
//! instead of the UI thread.
//!
//! The modal is EPHEMERAL (MOD-3.10): the analysis lives in
//! [`BeatifySession`] until Save writes the two artifacts, and dismissing
//! the modal leaves nothing behind. Reading corrections and the warp
//! slider re-query the session; neither re-runs the tracker (MOD-26) and
//! neither renders audio (MOD-A22).
//!
//! Serialization here is camelCase throughout, because the §5 payload the
//! tab emits is specified that way and one convention per feature beats
//! two.

use dj_analysis::beatify::{
    self, audition, detect, grid, store, Agreement, Analysis, Grid, Quality, Reading, Ruler,
    Sweep, WarpMap,
};
use dj_analysis::AudioData;
use dj_library::Track;
use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use std::sync::{Arc, Mutex};
use tauri::State;

use crate::{err, AppState, CmdError, CmdResult};

/// Waveform overview resolution cap (same law as the Clip page).
const MAX_BUCKETS: usize = 20_000;
/// Audition window cap: preview bytes cross the IPC boundary in one piece.
const MAX_PREVIEW_SECS: f64 = 120.0;

/// The modal's in-flight analysis. One at a time: the tab shows one track.
#[derive(Default)]
pub struct BeatifySession(Mutex<Option<Session>>);

struct Session {
    track_id: i64,
    source_path: PathBuf,
    source_hash: String,
    audio: Arc<AudioData>,
    analysis: Analysis,
}

impl BeatifySession {
    fn with<T>(&self, f: impl FnOnce(&mut Session) -> CmdResult<T>) -> CmdResult<T> {
        let mut guard = self
            .0
            .lock()
            .map_err(|_| CmdError::invalid("beatify: session lock poisoned"))?;
        match guard.as_mut() {
            Some(session) => f(session),
            None => Err(CmdError::invalid(
                "beatify: nothing analyzed yet — open a track first",
            )),
        }
    }

    fn set(&self, session: Session) -> CmdResult<()> {
        let mut guard = self
            .0
            .lock()
            .map_err(|_| CmdError::invalid("beatify: session lock poisoned"))?;
        *guard = Some(session);
        Ok(())
    }

    /// MOD-A25: cancel discards everything, because nothing was created.
    fn clear(&self) {
        if let Ok(mut guard) = self.0.lock() {
            *guard = None;
        }
    }
}

// ---------------------------------------------------------------------------
// Wire types
// ---------------------------------------------------------------------------

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BeatifySource {
    pub track_id: i64,
    pub title: String,
    pub artist: String,
    pub duration_secs: f64,
    pub sample_rate: u32,
    pub channels: usize,
    pub peaks: Vec<f32>,
}

/// Everything phase 1 and phase 2 draw, in one payload.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BeatifyAnalysis {
    pub source: BeatifySource,
    pub tracker: String,
    /// Analyzed span in source seconds — and therefore the import (MOD-A8).
    pub region: [f64; 2],
    pub grid: Grid,
    pub reading: Reading,
    pub agreement: Agreement,
    /// Detections in source seconds (amber: what was played).
    pub beats: Vec<f64>,
    pub confidence: Vec<f32>,
    pub drift: Vec<grid::DriftSpan>,
    pub sweep: Sweep,
    /// Slider position the sweep recommends.
    pub strength: f64,
    pub quality: Quality,
    /// Per-beat residuals at `strength`, seconds (the error strip).
    pub residuals: Vec<f64>,
    pub anchors: Vec<f64>,
    pub lead_in: f64,
    /// MOD-23 auto-flag: the interval histogram is bimodal at 2:1.
    pub metrical_flag: bool,
    pub output_secs: f64,
}

/// What the warp slider asks for on every move (MOD-A22): arithmetic only.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BeatifyMeters {
    pub strength: f64,
    /// Anchor spacing in beats; 0 is the no-warp position.
    pub anchor_stride: usize,
    pub quality: Quality,
    pub residuals: Vec<f64>,
    pub anchors: Vec<f64>,
}

/// A beatified track as the track view needs it.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BeatifyTrack {
    /// The project this render belongs to. Every clip command is keyed by
    /// it, because a track can have several.
    pub project_id: String,
    pub project_name: String,
    pub track_id: i64,
    pub title: String,
    pub artist: String,
    pub record: beatify::BeatifyRecord,
    pub duration_secs: f64,
    pub sample_rate: u32,
    pub channels: usize,
    pub peaks: Vec<f32>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveRequest {
    pub strength: f64,
    pub lead_in: f64,
    pub ruler_group: u32,
    /// The project to write into. Empty mints a new one — which is what
    /// "new project" does; re-beatifying passes the id it is replacing.
    #[serde(default)]
    pub project_id: String,
    /// What to call a newly minted project. Ignored for an existing one,
    /// which keeps the name it already has (rename is its own command).
    #[serde(default)]
    pub name: String,
}

/// A project as the tab's list shows it: the envelope, plus the facts
/// about the track and grid a person needs to tell two of them apart.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProjectSummary {
    pub id: String,
    pub name: String,
    pub track_id: i64,
    pub title: String,
    pub artist: String,
    pub bpm: f64,
    pub beats: usize,
    pub updated: u64,
    /// The source track is missing from the library (deleted, or a
    /// different machine): the project still opens, since its render is
    /// its own, but stems and re-beatify need the source.
    pub source_missing: bool,
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn describe(track: &Track, audio: &AudioData, buckets: usize) -> BeatifySource {
    BeatifySource {
        track_id: track.id,
        title: track.title.clone(),
        artist: track.artist.clone(),
        duration_secs: audio.duration_secs(),
        sample_rate: audio.sample_rate,
        channels: audio.channels.len(),
        peaks: dj_analysis::clip::peaks(audio, buckets.min(MAX_BUCKETS)),
    }
}

fn summarize(session: &Session, buckets: usize, track: &Track) -> BeatifyAnalysis {
    let a = &session.analysis;
    let strength = a.sweep.default_strength;
    BeatifyAnalysis {
        source: describe(track, &session.audio, buckets),
        tracker: a.tracker.clone(),
        region: a.region,
        grid: a.grid,
        reading: a.reading,
        agreement: a.agreement.clone(),
        beats: a.beats.clone(),
        confidence: a.confidence.clone(),
        drift: a.drift_spans(),
        sweep: a.sweep.clone(),
        strength,
        quality: a.quality_at(strength),
        residuals: a.residuals_at(strength),
        anchors: a.anchors_at(strength).iter().map(|x| x.dst).collect(),
        lead_in: a.lead_in,
        metrical_flag: a.metrical_flag,
        output_secs: a.output_secs(),
    }
}

fn wav(audio: &AudioData) -> CmdResult<tauri::ipc::Response> {
    if audio.frames() == 0 {
        return Err(CmdError::invalid("beatify: nothing to play there"));
    }
    Ok(tauri::ipc::Response::new(dj_analysis::clip::wav16_bytes(
        audio,
    )))
}

// ---------------------------------------------------------------------------
// Commands
// ---------------------------------------------------------------------------

/// Which beat tracker the tab will use, and how to install the good one.
///
/// `beat_this` is an optional runtime dependency (PyTorch): when it is not
/// importable the tab stays fully usable on the built-in DSP tracker and
/// shows the install hint, exactly like the YouTube provider without
/// `yt-dlp`.
#[tauri::command(async)]
pub fn beatify_tracker_status() -> detect::TrackerStatus {
    detect::tracker_status()
}

/// Detect beats and fit a grid. `region` is the import span (MOD-A8);
/// `None` is the whole file, which is what the modal runs on open
/// (MOD-A1/MOD-A11).
#[tauri::command(async)]
pub fn beatify_analyze(
    state: State<AppState>,
    track_id: i64,
    region: Option<[f64; 2]>,
    buckets: usize,
) -> CmdResult<BeatifyAnalysis> {
    let track = state.library.track(track_id).map_err(err)?;
    let path = PathBuf::from(&track.file_path);
    let audio = Arc::new(
        dj_analysis::decode_audio(&path).map_err(|e| err(format!("decoding {}: {e}", track.title)))?,
    );
    let tracker = detect::default_tracker();
    let analysis = beatify::analyze(
        &audio,
        tracker.as_ref(),
        region.map(|r| (r[0], r[1])),
        Reading::default(),
    )
    .map_err(|e| CmdError::invalid(e.to_string()))?;
    let session = Session {
        track_id,
        source_path: path,
        source_hash: track.content_hash.clone(),
        audio,
        analysis,
    };
    let summary = summarize(&session, buckets, &track);
    state.beatify.set(session)?;
    Ok(summary)
}

/// Re-read the current detections at another metrical level or phase
/// (MOD-23/MOD-24). Never re-runs the tracker.
#[tauri::command(async)]
pub fn beatify_set_reading(
    state: State<AppState>,
    reading: Reading,
    buckets: usize,
) -> CmdResult<BeatifyAnalysis> {
    let track_id = state.beatify.with(|s| Ok(s.track_id))?;
    let track = state.library.track(track_id).map_err(err)?;
    state.beatify.with(|session| {
        session.analysis = session
            .analysis
            .with_reading(reading)
            .map_err(|e| CmdError::invalid(e.to_string()))?;
        Ok(summarize(session, buckets, &track))
    })
}

/// Meters for a slider position (MOD-13): anchor arithmetic, no audio.
#[tauri::command(async)]
pub fn beatify_meters(state: State<AppState>, strength: f64) -> CmdResult<BeatifyMeters> {
    state.beatify.with(|session| {
        let a = &session.analysis;
        Ok(BeatifyMeters {
            strength,
            anchor_stride: grid::anchor_stride(strength).unwrap_or(0),
            quality: a.quality_at(strength),
            residuals: a.residuals_at(strength),
            anchors: a.anchors_at(strength).iter().map(|x| x.dst).collect(),
        })
    })
}

/// Audition a window (MOD-A20/A21/A23). `warped = false` plays the source
/// (phase 1), `true` renders just those beats through the warp (phase 2).
/// `click` mixes the metronome on top (MOD-27).
#[tauri::command(async)]
pub fn beatify_preview(
    state: State<AppState>,
    start_secs: f64,
    secs: f64,
    warped: bool,
    strength: f64,
    click: bool,
) -> CmdResult<tauri::ipc::Response> {
    let secs = secs.min(MAX_PREVIEW_SECS);
    state.beatify.with(|session| {
        let a = &session.analysis;
        let mut window = if warped {
            beatify::render_window(&session.audio, a, strength, start_secs, secs)
        } else {
            dj_analysis::clip::slice(&session.audio, start_secs, secs)
        };
        if click {
            let times: Vec<f64> = if warped {
                a.output_grid_times()
            } else {
                // Phase 1 ticks the DETECTIONS: over unwarped audio that is
                // what tells the user whether the metrical level and phase
                // are right, which a straight grid on a drifting song
                // cannot.
                a.beats.clone()
            };
            let local: Vec<f64> = times
                .iter()
                .map(|t| t - start_secs)
                .filter(|t| *t >= 0.0 && *t < secs)
                .collect();
            audition::mix_click(&mut window, &local, 0.5);
        }
        wav(&window)
    })
}

/// The sync check (MOD-28): four beats from each end of the track,
/// layered and looped.
#[tauri::command(async)]
pub fn beatify_sync_check(
    state: State<AppState>,
    strength: f64,
) -> CmdResult<tauri::ipc::Response> {
    state.beatify.with(|session| {
        let layered = audition::sync_check(&session.audio, &session.analysis, strength);
        wav(&layered)
    })
}

/// Commit (MOD-29/MOD-A24): render the warp once, write the artifacts
/// under `<data_dir>/beatify/<project>/` plus the sidecar, and hand the
/// builder its project.
///
/// An empty `project_id` MINTS a project; a given one overwrites that
/// project's grid and render (re-beatify), keeping its clips — which is
/// why the builder warns that they may no longer line up.
#[tauri::command(async)]
pub fn beatify_save(
    state: State<AppState>,
    request: SaveRequest,
    buckets: usize,
) -> CmdResult<BeatifyTrack> {
    let track_id = state.beatify.with(|s| Ok(s.track_id))?;
    let track = state.library.track(track_id).map_err(err)?;
    let data_dir = state.library.data_dir().to_path_buf();
    let existing = store::list(&data_dir).map_err(err)?;
    state.beatify.with(|session| {
        let (warped, map) = beatify::render(&session.audio, &session.analysis, request.strength);
        let warped_name = warped_name(&session.source_path);
        let record = beatify::record(
            &session.analysis,
            &beatify::Commit {
                source: &session.source_path,
                source_hash: &session.source_hash,
                warped_name: &warped_name,
                strength: request.strength,
                lead_in: request.lead_in.clamp(0.0, grid::LEAD_IN_MAX),
                ruler: Ruler {
                    group: request.ruler_group.max(1),
                },
            },
            &map,
        );
        let previous = if request.project_id.is_empty() {
            None
        } else {
            store::project(&data_dir, &request.project_id).map_err(err)?
        };
        let project = store::Project {
            id: previous
                .as_ref()
                .map(|p| p.id.clone())
                .unwrap_or_else(|| store::new_id(&existing)),
            name: previous
                .as_ref()
                .map(|p| p.name.clone())
                .unwrap_or_else(|| project_name(&request.name, &track)),
            track_id,
            source_hash: session.source_hash.clone(),
            updated: store::now_secs(),
        };
        store::save(&data_dir, &project, &record, &warped).map_err(err)?;
        Ok(BeatifyTrack {
            project_id: project.id,
            project_name: project.name,
            track_id,
            title: track.title.clone(),
            artist: track.artist.clone(),
            duration_secs: warped.duration_secs(),
            sample_rate: warped.sample_rate,
            channels: warped.channels.len(),
            peaks: dj_analysis::clip::peaks(&warped, buckets.min(MAX_BUCKETS)),
            record,
        })
    })
}

/// A new project is named for its track unless the user said otherwise.
fn project_name(asked: &str, track: &Track) -> String {
    let asked = asked.trim();
    if !asked.is_empty() {
        return asked.to_string();
    }
    if track.title.trim().is_empty() {
        format!("project {}", track.id)
    } else {
        track.title.clone()
    }
}

/// `boys.wav` → `boys.beatified.wav` (§3.11), display only: the file on
/// disk is always `warped.wav` inside the hash-keyed record directory.
fn warped_name(source: &std::path::Path) -> String {
    let stem = source
        .file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or("track");
    format!("{stem}.beatified.wav")
}

/// Every project in the store, newest first. This is the tab's front
/// door: a project, not a track, is what gets opened.
#[tauri::command(async)]
pub fn beatify_projects(state: State<AppState>) -> CmdResult<Vec<ProjectSummary>> {
    let data_dir = state.library.data_dir();
    let mut out = Vec::new();
    for project in store::list(data_dir).map_err(err)? {
        let Some(record) = store::load(data_dir, &project.id).map_err(err)? else {
            continue;
        };
        let track = resolve_track(&state, &project);
        out.push(ProjectSummary {
            id: project.id,
            name: project.name,
            track_id: track.as_ref().map(|t| t.id).unwrap_or(project.track_id),
            title: track
                .as_ref()
                .map(|t| t.title.clone())
                .unwrap_or_else(|| "(source missing)".into()),
            artist: track.as_ref().map(|t| t.artist.clone()).unwrap_or_default(),
            bpm: record.grid.bpm,
            beats: record.grid.beats,
            updated: project.updated,
            source_missing: track.is_none(),
        });
    }
    Ok(out)
}

/// The library row a project came from: by hash first, because ids are
/// re-assigned when a track is re-imported and the audio is not.
fn resolve_track(state: &AppState, project: &store::Project) -> Option<Track> {
    if let Ok(Some(found)) = state.library.track_by_hash(&project.source_hash) {
        return Some(found);
    }
    state.library.track(project.track_id).ok()
}

/// Open a project: its record and its render, ready for the builder.
#[tauri::command(async)]
pub fn beatify_project_open(
    state: State<AppState>,
    project_id: String,
    buckets: usize,
) -> CmdResult<Option<BeatifyTrack>> {
    let data_dir = state.library.data_dir();
    let (Some(project), Some(record)) = (
        store::project(data_dir, &project_id).map_err(err)?,
        store::load(data_dir, &project_id).map_err(err)?,
    ) else {
        return Ok(None);
    };
    let track = resolve_track(&state, &project);
    let path = store::warped_path(data_dir, &project_id);
    let warped = dj_analysis::decode_audio(&path)
        .map_err(|e| err(format!("reading the beatified render: {e}")))?;
    Ok(Some(BeatifyTrack {
        project_id: project.id,
        project_name: project.name,
        track_id: track.as_ref().map(|t| t.id).unwrap_or(project.track_id),
        title: track
            .as_ref()
            .map(|t| t.title.clone())
            .unwrap_or_else(|| "(source missing)".into()),
        artist: track.map(|t| t.artist).unwrap_or_default(),
        duration_secs: warped.duration_secs(),
        sample_rate: warped.sample_rate,
        channels: warped.channels.len(),
        peaks: dj_analysis::clip::peaks(&warped, buckets.min(MAX_BUCKETS)),
        record,
    }))
}

/// Rename a project. The name is a label, nothing keys off it.
#[tauri::command(async)]
pub fn beatify_project_rename(
    state: State<AppState>,
    project_id: String,
    name: String,
) -> CmdResult<Vec<ProjectSummary>> {
    let data_dir = state.library.data_dir();
    let mut project = store::project(data_dir, &project_id)
        .map_err(err)?
        .ok_or_else(|| CmdError::invalid(format!("beatify: no project {project_id}")))?;
    let name = name.trim();
    if name.is_empty() {
        return Err(CmdError::invalid("beatify: a project needs a name"));
    }
    project.name = name.to_string();
    project.updated = store::now_secs();
    store::save_project(data_dir, &project).map_err(err)?;
    beatify_projects(state)
}

/// Delete a project: its grid, its render, its clips, its warped stems.
/// The source track and the library are untouched.
#[tauri::command(async)]
pub fn beatify_project_delete(
    state: State<AppState>,
    project_id: String,
) -> CmdResult<Vec<ProjectSummary>> {
    store::remove(state.library.data_dir(), &project_id).map_err(err)?;
    beatify_projects(state)
}

/// A window of a project's render, for the track view's transport.
#[tauri::command(async)]
pub fn beatify_project_audio(
    state: State<AppState>,
    project_id: String,
    start_secs: f64,
    secs: f64,
) -> CmdResult<tauri::ipc::Response> {
    let path = store::warped_path(state.library.data_dir(), &project_id);
    let warped = dj_analysis::decode_audio(&path)
        .map_err(|e| err(format!("reading the beatified render: {e}")))?;
    wav(&dj_analysis::clip::slice(
        &warped,
        start_secs,
        secs.min(MAX_PREVIEW_SECS),
    ))
}

/// Dismiss the modal (MOD-A25): drop the session, write nothing.
#[tauri::command(async)]
pub fn beatify_cancel(state: State<AppState>) {
    state.beatify.clear();
}

/// The anchor map of the current session at a slider position — handy for
/// debugging and used by the modal's "what will Save write" readout.
#[tauri::command(async)]
pub fn beatify_warp_map(state: State<AppState>, strength: f64) -> CmdResult<Vec<[f64; 2]>> {
    state
        .beatify
        .with(|session| Ok(WarpMap::from_anchors(&session.analysis.anchors_at(strength)).pairs()))
}

File: app/src/beatify.ts
// Beatify tab bridge: types mirror `dj_analysis::beatify` (camelCase on
// both sides, because the §5 payload the tab emits is specified that way)
// and the grid math below is pure, so the page's behaviour is testable
// without a backend.
//
// Beat n is at `phase + n * period`. Exactly. There is no beat array, no
// meter and no bars anywhere in this file: `ruler.group` is a display
// preference and nothing else reads it.

import { IpcClient } from './ipc';

export interface Grid {
  bpm: number;
  period: number;
  /** Seconds of beat 0 in the warped file — always one period. */
  phase: number;
  beats: number;
}

export interface Reading {
  /** Metrical level: 2 doubles the tempo, 0.5 halves it. */
  factor: number;
  half_shift?: boolean;
  halfShift?: boolean;
}

export type Verdict = 'unanimous' | 'mostlyAgreed' | 'split' | 'metricalSplit' | 'singleTracker';

export interface SeedReading {
  seed: string;
  bpm: number;
  beats: number;
}

export interface Agreement {
  verdict: Verdict;
  tempoSpreadBpm: number;
  phaseAgreementPct: number;
  metricalSplit: boolean;
  readings: SeedReading[];
  disagreementSpans: [number, number][];
}

export interface Quality {
  worstFlamMs: number;
  peakStretchPct: number;
  rmsMs: number;
  inBandPct: number;
}

export interface SweepPoint {
  strength: number;
  stride: number;
  quality: Quality;
}

export interface Sweep {
  points: SweepPoint[];
  zone: [number, number] | null;
  defaultStrength: number;
}

export interface DriftSpan {
  startSecs: number;
  endSecs: number;
  deltaBpm: number;
}

export interface BeatifySource {
  trackId: number;
  title: string;
  artist: string;
  durationSecs: number;
  sampleRate: number;
  channels: number;
  peaks: number[];
}

export interface BeatifyAnalysis {
  source: BeatifySource;
  tracker: string;
  region: [number, number];
  grid: Grid;
  reading: Reading;
  agreement: Agreement;
  beats: number[];
  confidence: number[];
  drift: DriftSpan[];
  sweep: Sweep;
  strength: number;
  quality: Quality;
  residuals: number[];
  anchors: number[];
  leadIn: number;
  metricalFlag: boolean;
  outputSecs: number;
}

export interface BeatifyMeters {
  strength: number;
  anchorStride: number;
  quality: Quality;
  residuals: number[];
  anchors: number[];
}

export interface BeatifyRecord {
  source: string;
  sourceHash: string;
  sourceSpan: [number, number];
  warped: string;
  grid: Grid;
  leadIn: number;
  ruler: { group: number };
  warp: { strength: number; anchorStride: number; map: [number, number][] };
  quality: Quality;
  analysis: { tracker: string; agreement: Agreement; confidence: number[] };
  reading: Reading;
}

/** A beatified take on one track: its grid, its render, its clips. One
 *  track can have any number of them, so the PROJECT id — not the track
 *  id — is what every clip command is keyed by. */
export interface BeatifyProject {
  id: string;
  name: string;
  trackId: number;
  title: string;
  artist: string;
  bpm: number;
  beats: number;
  /** Unix seconds; the list is newest first. */
  updated: number;
  /** The source track has left the library: the project still opens and
   *  plays, but stems and re-beatify need the original file. */
  sourceMissing: boolean;
}

export interface BeatifyTrack {
  projectId: string;
  projectName: string;
  trackId: number;
  title: string;
  artist: string;
  record: BeatifyRecord;
  durationSecs: number;
  sampleRate: number;
  channels: number;
  peaks: number[];
}

export interface TrackerStatus {
  tracker: string;
  beatThis: boolean;
  seeds: string[];
  device: string;
  python: string;
  detail: string;
  installHint: string;
}

// ---------------------------------------------------------------------------
// Grid math (OUT-1: three numbers describe the whole rhythmic content)
// ---------------------------------------------------------------------------

export const DEFAULT_RULER_GROUP = 4;
/** Residual band the user cannot hear — mirrors `grid::IN_BAND_SECS`. */
export const IN_BAND_MS = 5;
/** Meter thresholds — mirror `grid::FLAM_GREEN_MS` / `STRETCH_GREEN_PCT`. */
export const FLAM_GREEN_MS = 5;
export const STRETCH_GREEN_PCT = 1.2;
/** Lead-in range in milliseconds (MOD-20). */
export const LEAD_IN_MAX_MS = 40;

export function beatTime(grid: Grid, n: number): number {
  return grid.phase + n * grid.period;
}

/** Which beat a time is nearest to (TV-6: beats are the atomic unit). */
export function beatAt(grid: Grid, secs: number): number {
  if (grid.period <= 0) return 0;
  return Math.round((secs - grid.phase) / grid.period);
}

export function clampBeat(grid: Grid, n: number): number {
  return Math.min(grid.beats - 1, Math.max(0, n));
}

/** Time of the nearest beat, clamped into the track. */
export function snapTime(grid: Grid, secs: number): number {
  return beatTime(grid, clampBeat(grid, beatAt(grid, secs)));
}

/** TV-14: selections snap OUTWARD, so they are always whole beats. */
export function snapSelection(
  grid: Grid,
  fromSecs: number,
  toSecs: number,
): { startBeat: number; endBeat: number } {
  const lo = Math.min(fromSecs, toSecs);
  const hi = Math.max(fromSecs, toSecs);
  const startBeat = clampBeat(grid, Math.floor((lo - grid.phase) / grid.period));
  const endBeat = clampBeat(grid, Math.ceil((hi - grid.phase) / grid.period));
  return { startBeat, endBeat: Math.max(endBeat, startBeat + 1) };
}

/** TV-15: "12 beats · 3 groups" when the count divides evenly. */
export function selectionLabel(beats: number, group: number): string {
  const plural = beats === 1 ? 'beat' : 'beats';
  if (group > 1 && beats % group === 0) {
    const groups = beats / group;
    return `${beats} ${plural} · ${groups} ${groups === 1 ? 'group' : 'groups'}`;
  }
  return `${beats} ${plural}`;
}

/** Grid line times over a beat range, every `step` beats (TV-1: pure
 *  arithmetic — the warped audio needs no per-beat storage). */
export function gridLines(grid: Grid, fromBeat: number, toBeat: number, step: number): number[] {
  const out: number[] = [];
  if (step <= 0) return out;
  const first = Math.max(0, Math.ceil(fromBeat / step) * step);
  const last = Math.min(toBeat, grid.beats - 1);
  for (let n = first; n <= last; n += step) out.push(beatTime(grid, n));
  return out;
}

/** Zoom levels in visible beats (TV-10); `null` is "whole track". */
export const ZOOM_BEATS: (number | null)[] = [null, 512, 128, 32, 8, 2, 1];

export interface GridLod {
  /** Spacing of drawn lines, in beats. */
  step: number;
  /** Spacing of emphasized lines, in beats. */
  emphasis: number;
  /** Density band from per-beat confidence (TV-5). */
  density: boolean;
  subdivisions: boolean;
}

/** TV-2: what to draw at a given zoom. Emphasis follows the ruler
 *  grouping, never any detected structure — there isn't any. */
export function gridLod(visibleBeats: number, group: number): GridLod {
  const g = Math.max(1, Math.round(group));
  if (visibleBeats > 256) {
    return { step: 16 * g, emphasis: 16 * g, density: false, subdivisions: false };
  }
  if (visibleBeats > 64) {
    return { step: 4 * g, emphasis: 16 * g, density: false, subdivisions: false };
  }
  if (visibleBeats > 16) {
    return { step: g, emphasis: 4 * g, density: false, subdivisions: false };
  }
  if (visibleBeats >= 4) {
    return { step: 1, emphasis: g, density: false, subdivisions: false };
  }
  return { step: 1, emphasis: g, density: true, subdivisions: true };
}

/** TV-25: a loop changed while the playhead is outside it wraps at the
 *  next GROUP boundary — immediate jumping is jarring, waiting for a loop
 *  end the playhead may never reach is a hang. (§6 open question 1: group,
 *  the recommended option.) */
export function loopWrapBeat(playheadBeat: number, group: number): number {
  const g = Math.max(1, Math.round(group));
  return Math.ceil((playheadBeat + 1e-9) / g) * g;
}

/** Anchor spacing for a slider position — mirrors `grid::anchor_stride`. */
export function anchorStride(strength: number): number {
  const s = Math.min(1, Math.max(0, strength));
  if (s <= 0.02) return 0;
  const stride = Math.round(2 ** (Math.log2(64) * (1 - s)));
  return Math.min(64, Math.max(1, stride));
}

export function verdictLabel(agreement: Agreement): string {
  switch (agreement.verdict) {
    case 'unanimous':
      return 'UNANIMOUS';
    case 'mostlyAgreed':
      return 'MOSTLY AGREED';
    case 'metricalSplit':
      return 'METRICAL SPLIT';
    case 'split':
      return 'SPLIT';
    default:
      return 'SINGLE TRACKER';
  }
}

/** Green / amber / red for the verdict line (MOD-30). */
export function qualityLevel(q: Quality): 'good' | 'warn' | 'bad' {
  if (q.worstFlamMs < FLAM_GREEN_MS && q.peakStretchPct < STRETCH_GREEN_PCT) return 'good';
  if (q.worstFlamMs < FLAM_GREEN_MS * 3) return 'warn';
  return 'bad';
}

export function readingOf(reading: Reading | undefined): {
  factor: number;
  halfShift: boolean;
} {
  return {
    factor: reading?.factor ?? 1,
    halfShift: reading?.halfShift ?? reading?.half_shift ?? false,
  };
}

export function timecode(secs: number): string {
  if (!Number.isFinite(secs) || secs < 0) return '0:00.00';
  const m = Math.floor(secs / 60);
  const s = secs - m * 60;
  return `${m}:${s.toFixed(2).padStart(5, '0')}`;
}

// ---------------------------------------------------------------------------
// IPC
// ---------------------------------------------------------------------------

/** What Save commits. An empty `projectId` mints a project (that is what
 *  "new project" sends); re-beatifying passes the id it is replacing. */
export interface SaveRequest {
  strength: number;
  leadIn: number;
  rulerGroup: number;
  projectId: string;
  name: string;
}

/** What the Beatify page needs; tests substitute a mock. */
export interface BeatifyClientApi {
  trackerStatus(): Promise<TrackerStatus | null>;
  analyze(
    trackId: number,
    region: [number, number] | null,
    buckets: number,
  ): Promise<BeatifyAnalysis | null>;
  setReading(reading: Reading, buckets: number): Promise<BeatifyAnalysis | null>;
  meters(strength: number): Promise<BeatifyMeters | null>;
  preview(
    startSecs: number,
    secs: number,
    warped: boolean,
    strength: number,
    click: boolean,
  ): Promise<ArrayBuffer | null>;
  syncCheck(strength: number): Promise<ArrayBuffer | null>;
  save(request: SaveRequest, buckets: number): Promise<BeatifyTrack | null>;
  projects(): Promise<BeatifyProject[] | null>;
  openProject(projectId: string, buckets: number): Promise<BeatifyTrack | null>;
  renameProject(projectId: string, name: string): Promise<BeatifyProject[] | null>;
  deleteProject(projectId: string): Promise<BeatifyProject[] | null>;
  projectAudio(projectId: string, startSecs: number, secs: number): Promise<ArrayBuffer | null>;
  cancel(): Promise<void>;
}

export class BeatifyClient extends IpcClient implements BeatifyClientApi {
  trackerStatus() {
    return this.call<TrackerStatus>('beatify_tracker_status');
  }
  analyze(trackId: number, region: [number, number] | null, buckets: number) {
    return this.call<BeatifyAnalysis>('beatify_analyze', { trackId, region, buckets });
  }
  setReading(reading: Reading, buckets: number) {
    return this.call<BeatifyAnalysis>('beatify_set_reading', { reading, buckets });
  }
  meters(strength: number) {
    return this.call<BeatifyMeters>('beatify_meters', { strength });
  }
  preview(startSecs: number, secs: number, warped: boolean, strength: number, click: boolean) {
    return this.call<ArrayBuffer>('beatify_preview', {
      startSecs,
      secs,
      warped,
      strength,
      click,
    });
  }
  syncCheck(strength: number) {
    return this.call<ArrayBuffer>('beatify_sync_check', { strength });
  }
  save(request: SaveRequest, buckets: number) {
    return this.call<BeatifyTrack>('beatify_save', { request, buckets });
  }
  projects() {
    return this.call<BeatifyProject[]>('beatify_projects');
  }
  openProject(projectId: string, buckets: number) {
    return this.call<BeatifyTrack>('beatify_project_open', { projectId, buckets });
  }
  renameProject(projectId: string, name: string) {
    return this.call<BeatifyProject[]>('beatify_project_rename', { projectId, name });
  }
  deleteProject(projectId: string) {
    return this.call<BeatifyProject[]>('beatify_project_delete', { projectId });
  }
  projectAudio(projectId: string, startSecs: number, secs: number) {
    return this.call<ArrayBuffer>('beatify_project_audio', { projectId, startSecs, secs });
  }
  async cancel() {
    await this.call<null>('beatify_cancel');
  }
}

export const beatifyClient = new BeatifyClient();
